pytorch_lightning/trainer/supporters.py

```diff
@@ -176,3 +181,316 @@ def to_disk(self) -> None:
+        self.counter = 0
+
+    def __iter__(self) -> Any:
+        """
+
+        Creates the internal iterator and returns self
+
+        Returns:
+            CycleIterator: self
+
+        """
+        self.counter = 0
+        self._loader_iter = iter(self.loader)
+        return self
+
+    def __next__(self) -> Any:
+        """
+        Fetches the next batch from internal dataloader and restarts
+        it if necessary
+
+        Returns:
+            Any: the resulting batch
+
+        Raises:
+            StopIteration: if more then :attr:`length` batches have been returned
+
+        """
+        # Note: if self.length is `inf`, then the iterator will never stop
+        if self.counter >= self.__len__():
+            raise StopIteration
+
+        try:
+            return next(self._loader_iter)
+
+        except StopIteration:
+            self._loader_iter = iter(self.loader)
+            return next(self._loader_iter)
+
+        finally:
+            self.counter += 1
+
+    def __len__(self) -> Union[int, float]:
+        return self.length
+
+
+class CombinedDataset(object):
+    """
+    Combine multiple datasets and compute their statistics
+    """
+    def __init__(self, datasets: Union[Sequence, Mapping]):
+        """
+
+        Args:
+            datasets: a sequence/mapping datasets. Can be a collections of torch.utils.Dataset,
+                Iterable or even None.
+
+        """
+        self.datasets = datasets
+
+        self.max_len = self._calc_num_data(self.datasets, 'max')
+        self.min_len = self._calc_num_data(self.datasets, 'min')
+
+    @staticmethod
+    def _calc_num_data(datasets: Union[Sequence, Mapping], mode: str = 'min') -> Union[int, float]:
+        """
+        Compute the length of `CombinedDataset` according to the `mode`.
+
+        Args:
+            datasets: a sequence/mapping datasets. Can be a collections of torch.utils.data.Dataset,
+                Iterable or even None.
+            mode (str): 'min' or 'max'. Determine `CombinedDataset`'s length is the maximum or minimum of
+                the datasets.
+
+        Returns:
+            length (int): the length of `CombinedDataset`
+
+        """
+        if mode not in ['min', 'max']:
+            raise ValueError(f"Invalid Mode: {mode}")
+
+        # extract the lengths
+        all_lengths = apply_to_collection(datasets, (Dataset, Iterable, type(None)), get_len,
+                                          wrong_dtype=(Sequence, Mapping))
+
+        compute_func = eval(mode)
+
+        if isinstance(all_lengths, (int, float)):
+            length = all_lengths
+
+        elif isinstance(all_lengths, Mapping):
+            length = compute_func(all_lengths.values())
+
+        elif isinstance(all_lengths, Sequence):
+            length = compute_func(all_lengths)
+
+        return length
+
+    def __len__(self) -> int:
+        """Return the minimum length of the datasets."""
+        return self.min_len
```

Comment:
should this not depend on the mode?

Reply:
Hey @justusschock. Can you check this one ?

Reply:
done